hwp_gripper/control/pru_monitor.py

```diff
@@ -0,0 +1,291 @@
+        UDP packet data
+    """
+    header = struct.unpack("H", data[:2])
+    if header not in packet_headers:
+        raise RuntimeError(f"Bad header: {header}")
+
+    packet_type = packet_headers[header]
+    return packet_type.from_data(data[2:])
+
+
+########################################
+# Gripper state definition
+########################################
+
+@dataclass
+class LimitState:
+    pru_bit: Optional[int] = None
+    state: bool = False
+
+
+@dataclass
+class ActuatorState:
+    axis: int
+    limit_pru_bits: Tuple[int, int]
+
+    limits: Optional[Dict[str, LimitState]] = {
+        'cold_grip': LimitState(),
+        'warm_grip': LimitState(),
+    }
+    pos: float = 0
+    calibrated: bool = False
+
+    def __post_init__(self):
+        self.limits['cold_grip'].pru_bit = self.limit_pru_bits[0]
+        self.limits['warm_grip'].pru_bit = self.limit_pru_bits[1]
+
+
+@dataclass
+class GripperState:
+    actuators: Tuple[ActuatorState, ActuatorState, ActuatorState] = (
+        ActuatorState(axis=1, limit_pru_bits=(8, 9)),
+        ActuatorState(axis=2, limit_pru_bits=(10, 11)),
+        ActuatorState(axis=3, limit_pru_bits=(12, 13)),
+    )
+
+    last_packet_received: float = 0.0
+    last_limit_received: float = 0.0
+    last_encoder_received: float = 0.0
+
+    _expiration_time: float = 10.0
+    _last_enc_state: Optional[int] = None
+
+    @property
+    def expired(self):
+        """Returns True if the gripper state has not been updated in a while"""
+        return time.time() - self.last_packet_received > self._expiration_time
+
+    def update(self, packet):
+        """Updates the gripper state based on an incoming PRU packet"""
+        self.last_packet_received = packet.timestamp
+
+        if isinstance(packet, EncoderPacket):
+            # Update the position of each encoder
+            for act in self.actuators:
+                idx = act.axis - 1
+                a_state = (packet.state >> 2*idx) & 1
+                b_state = (packet.state >> (2*idx + 1)) & 1
+
+                if self._last_enc_state is not None:
+                    rising_edges = np.diff(
+                        a_state, prepend=(self._last_enc_state >> 2*idx) & 1
+                    ) == 1
+                    dirs = (b_state[rising_edges] * 2 - 1)
+                else:
+                    rising_edges = np.diff(a_state) == 1
+                    dirs = (b_state[1:][rising_edges] * 2 - 1)
+
+                act.pos += np.sum(dirs) * MM_PER_NOTCH
+
+            self._last_enc_state = packet.state[-1]
+
+        elif isinstance(packet, LimitPacket):
+            self.last_limit_received = packet.timestamp
+
+            # Update the limit state of each actuator
+            for act in self.actuators:
+                for limit in act.limits.values():
+                    limit.state = bool((packet.state >> limit.pru_bit) & 1)
+
+        elif isinstance(packet, ErrorPacket):
+            # Should we do anything else for errors?
```

Comment:
Looking more closely at the pru C code, it looks like these error packets are never generated. I adapted the gripper C code from the already existing hwp encoder C code and in that code an error packet was generated whenever there was a timing error in the incoming IRIG-B signal. The gripper does not have a timing signal, hence there is no need for the error packets. I should probably remove the unused code from BeagleboneGripper.c to make it less confusing.